S4TF_Tutorial_3.ipynb

```diff
@@ -0,0 +1,1768 @@
+            "text/plain": [
+              "<Figure size 288x288 with 2 Axes>"
+            ]
+          },
+          "metadata": {
+            "tags": []
+          }
+        },
+        {
+          "output_type": "execute_result",
+          "data": {
+            "text/plain": [
+              "None\n"
+            ]
+          },
+          "metadata": {
+            "tags": []
+          },
+          "execution_count": 10
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "liZMNGVSCLgd",
+        "colab_type": "text"
+      },
+      "source": [
+        "Display the first 25 images from the *training set* and display the class name below each image. Verify that the data is in the correct format and we're ready to build and train the network."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "RHBungPfCLrg",
+        "colab_type": "code",
+        "outputId": "28c588e1-becc-4c78-af81-df5ac6e4da40",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 768
+        }
+      },
+      "source": [
+        "plt.figure(figsize: [12, 12])\n",
+        "var dis = 0\n",
+        "for image in 0...24 {\n",
+        "    let TrainImage = trainImages[image]\n",
+        "    let TrainLabel = Int((trainNumericLabels[image].scalars)[0])\n",
```

Comment:
Every use of `.scalar` copies the entire tensor from the accelerator to the host (CPU) side. So it's better to avoid this pattern. Instead, do indexing on the tensor side and use `.scalarized()` to turn it to a scalar.
```suggestion
        "    let trainLabel = Int((trainNumericLabels[i][0].scalarized())\n",
```

Reply:
This gives me an error:

`Fatal error: Index out of range using input dim 0; input has only 0 dims: file /swift-base/swift/stdlib/public/TensorFlow/CompilerRuntime.swift, line 2123`

Reply:
The `trainNumericLabels[image][0]` itself has just 1 value. So whenever I .scalar it, it only copies 1 value, the same it will copy in case of `scalarized()`
I've done `Int((trainNumericLabels[image].scalars)[0])` because Int needs a binaryInteger that's why `[0]` index is used at the end so that it can tell the Int typecasting outside the bracket that only 1 value is being passed on to be converted to Int

Reply:
Ah right. In that case, `trainNumericLabels[i].scalarized()` should just work.

Reply:
Should I consider this as resolved?

Reply:
So, anything else left?